File: tools/polynomial_fitting/polynomial_fitting.cpp
#include "polynomial_fitting.h"

#include <iostream>

PolynomialFitting::PolynomialFitting(const std::vector<double>& x,
                                     const std::vector<double>& y, int degree)
    : x_data(x), y_data(y), degree(degree) {
  int n = x_data.size();
  vandermonde_matrix = Eigen::MatrixXd(n, degree + 1);

  // 构建 Vandermonde 矩阵
  for (int i = 0; i < n; ++i) {
    for (int j = 0; j <= degree; ++j) {
      vandermonde_matrix(i, j) = std::pow(x_data[i], j);
    }
  }
}

Eigen::VectorXd PolynomialFitting::FitPolynomial() {
  Eigen::VectorXd y_vec(y_data.size());
  for (size_t i = 0; i < y_data.size(); ++i) {
    y_vec(i) = y_data[i];
  }

  // 使用最小二乘法求解 (V^T * V) * coefficients = V^T * y
  coefficients = (vandermonde_matrix.transpose() * vandermonde_matrix)
                     .ldlt()
                     .solve(vandermonde_matrix.transpose() * y_vec);

  return coefficients;
}

double PolynomialFitting::Evaluate(double x) {
  double result = 0.0;
  for (int i = 0; i <= degree; ++i) {
    result += coefficients(i) * std::pow(x, i);
  }
  return result;
}

File: tools/polynomial_fitting/polynomial_fitting.py
import numpy as np
import matplotlib.pyplot as plt


class PolynomialFitting:
    def __init__(self, x_data, y_data, degree):
        self.x_data = np.array(x_data)
        self.y_data = np.array(y_data)
        self.degree = degree
        self.coefficients = None

    def fit_polynomial(self):
        # 使用 numpy.polyfit 进行多项式拟合，degree 表示多项式的阶数
        self.coefficients = np.polyfit(self.x_data, self.y_data, self.degree)
        return self.coefficients

    def evaluate(self, x):
        # 使用拟合的多项式系数来计算给定 x 值的多项式值
        poly = np.poly1d(self.coefficients)
        return poly(x)


if __name__ == "__main__":
    x_data = [1.0, 2.0, 3.0, 4.0, 5.0]
    y_data = [1.2, 2.8, 3.6, 5.0, 7.2]
    degree = 2

    pf = PolynomialFitting(x_data, y_data, degree)
    coefficients = pf.fit_polynomial()
    print("Fitted polynomial coefficients:")
    print(coefficients)
    for x in np.arange(1.0, 6.0, 1.0):
        print(f"P({x}) = {pf.evaluate(x)}")

    # 可视化拟合结果
    x_fit = np.linspace(1, 5, 100)
    y_fit = pf.evaluate(x_fit)
    plt.scatter(x_data, y_data, color="blue", label="Data points")
    plt.plot(x_fit, y_fit, color="red", label=f"Fitted Polynomial (degree={degree})")
    plt.xlabel("x")
    plt.ylabel("y")
    plt.legend()
    plt.title("Polynomial Fitting")
    plt.show()
